src/newgrf_profiling.cpp

```diff
@@ -156,17 +156,29 @@ uint32 NewGRFProfiler::FinishAll()
 		IConsolePrint(CC_DEBUG, "Total NewGRF callback processing: {} microseconds over {} ticks.", total_microseconds, max_ticks);
 	}
 
-	_newgrf_profile_end_date = MAX_DATE;
-
 	return total_microseconds;
 }
 
 /**
  * Check whether profiling is active and should be finished.
  */
-static IntervalTimer<TimerGameCalendar> _check_profiling_finished({TimerGameCalendar::DAY, TimerGameCalendar::Priority::NONE}, [](auto)
+static TimeoutTimer<TimerGameTick> _profiling_finish_timeout(0, []()
 {
-	if (_newgrf_profilers.empty() || _newgrf_profile_end_date > TimerGameCalendar::date) return;
-
 	NewGRFProfiler::FinishAll();
 });
+
+/**
+ * Start the timeout timer that will finish all profiling sessions.
+ */
+/* static */ void NewGRFProfiler::StartTimer(uint64 ticks)
+{
+	_profiling_finish_timeout.Reset(ticks);
+}
+
+/**
+ * Abort the timeout timer, so the timer callback is never called.
+ */
+/* static */ void NewGRFProfiler::AbortTimer()
+{
+	_profiling_finish_timeout.Abort();
+}
```

Comment:
I did consider just making `_profiling_finish_timeout` public, and letting the console commands call them directly. Saves a few wrapper functions. On the other hand, nothing wrong with wrapper functions.